File: projects/05/clojure/src/quadratic.clj
;;;; quadratic.clj finds the real roots of a quadratic function
;;;;
;;;; Input: a, b, and c from the quadratic equation
;;;; Output: real roots of quadratic
;;;;
;;;; Usage: clojure -m quadratic
;;;;
;;;; Completed by: Sarah Whitten
;;;; Date: March 5, 2019
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(ns quadratic)      ; name the program

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn quadratic [a b c]
	(if (not= a 0)
		(let 
			[ det (- (* b b) (* 4 a c))]
			(if (>= det 0)
				(let
					[ sqrtDet (Math/sqrt det)
					  root1 (/ (+ (- b) sqrtDet) (* 2 a))
					  root2 (/ (- (- b) sqrtDet) (* 2 a))
					]
					(vector root1 root2)
					; true
				)
				(println "\n*** quadraticRoots(): b^2 - 4ac is negative!")
				; false
			)
		)
		(println "\n*** QuadraticRoots(): a is zero!")
		; false
	)
)

(defn -main []
	(println "Enter a, b, and c from your quadratic function:")
	(let 
		[ a (float (read))
		  b (float (read))
		  c (float (read))
		  roots (quadratic a b c)
		  root1 (get roots 0)
		  root2 (get roots 1)
		]
		; if
		(printf (str "\nThe first root is %f and" 
                    " the second root is %f\n")
                       root1 root2)
	)
)

File: projects/05/java/Quadratic.java
/* Quadratic.java is a driver for function quadratic().
 * Author: Sarah Whitten
 * Date: March 4, 2019
 ****************************************************************/

import java.util.*;

public class Quadratic {

	public static void main(String[] args) {

		Scanner s = new Scanner(System.in);
    	System.out.print("Enter a, b, and c from your quadratic equation:\n");

    	int a = s.nextInt();
    	int b = s.nextInt();
    	int c = s.nextInt();
    	double[] roots = {0.0, 0.0};

    	if (quadratic(a, b, c, roots)) {
    		System.out.print("The real roots are: ");
    		System.out.print(roots[0]);
    		System.out.print(" and ");
    		System.out.print(roots[1]);
    		System.out.println();
    	}
	}

	public static boolean quadratic(int a, int b, int c, double[] roots) {

		double root1, root2;

		if (a != 0) {
        	double det = Math.pow(b, 2.0) - 4 * a * c;
         	if (det >= 0) {
            	roots[0] = (-b + Math.sqrt(det))/(2*a);
            	roots[1] = (-b - Math.sqrt(det))/(2*a);
            	return true;
         	} else {
            	System.err.println("\n*** quadraticRoots(): b^2 - 4ac is negative!");
            	roots[0] = roots[1] = 0.0;
            	return false;
         	}
      	} else {
         	System.err.println("\n*** QuadraticRoots(): a is zero!");
         	roots[0] = roots[1] = 0.0;
         	return false;
      }
	}
}
